core/domain/stats_domain.py

```diff
@@ -1351,3 +1365,366 @@ def validate(self):
+                learner_answer_info_list in bytes.
+            schema_version: int. The schema version of the
+                LearnerAnswerInfo class.
+        """
+
+        self.state_reference = state_reference
+        self.entity_type = entity_type
+        self.interaction_id = interaction_id
+        self.learner_answer_info_list = learner_answer_info_list
+        self.accumulated_answer_info_json_size_bytes = (
+            accumulated_answer_info_json_size_bytes)
+        self.schema_version = schema_version
+
+    def to_dict(self):
+        """Returns a dict representing LearnerAnswerDetails domain object.
+
+        Returns:
+            dict. A dict, mapping all fields of LearnerAnswerDetails instance.
+        """
+        return {
+            'state_reference': self.state_reference,
+            'entity_type': self.entity_type,
+            'interaction_id': self.interaction_id,
+            'learner_answer_info_list': [
+                learner_answer_info.to_dict() for learner_answer_info in (
+                    self.learner_answer_info_list)
+            ],
+            'accumulated_answer_info_json_size_bytes': (
+                self.accumulated_answer_info_json_size_bytes),
+            'schema_version': self.schema_version
+        }
+
+    @classmethod
+    def from_dict(cls, learner_answer_details_dict):
+        """Return a LearnerAnswerDetails domain object from a dict.
+
+        Args:
+            learner_answer_details_dict: dict. The dict representation of
+                LearnerAnswerDetails object.
+
+        Returns:
+            LearnerAnswerDetails. The corresponding LearnerAnswerDetails
+                domain object.
+        """
+        return cls(
+            learner_answer_details_dict['state_reference'],
+            learner_answer_details_dict['entity_type'],
+            learner_answer_details_dict['interaction_id'],
+            [LearnerAnswerInfo.from_dict(learner_answer_info_dict)
+             for learner_answer_info_dict in learner_answer_details_dict[
+                 'learner_answer_info_list']],
+            learner_answer_details_dict[
+                'accumulated_answer_info_json_size_bytes'],
+            learner_answer_details_dict['schema_version']
+        )
+
+    def validate(self):
+        """Validates LearnerAnswerDetails domain object."""
+
+        if not isinstance(self.state_reference, basestring):
+            raise utils.ValidationError(
+                'Expected state_reference to be a string, received %s' % str(
+                    self.state_reference))
+
+        if not isinstance(self.entity_type, basestring):
+            raise utils.ValidationError(
+                'Expected entity_type to be a string, received %s' % str(
+                    self.entity_type))
+
+        split_state_reference = self.state_reference.split('.')
+        if self.entity_type == feconf.ENTITY_TYPE_EXPLORATION:
+            if len(split_state_reference) < 2:
```

Comment:
Not equal check can't be used here as state_reference for entity type exploration will have at least one ```.``` in it, so after splitting there should be at least 2 elements in the list.
While in case of question id it's in a hash, therefore, its state reference will not split hence the list will contain only one element.

Reply:
Like for state name which have punctuation in it like 'intro.state' in this case, it can be more than 2.

Reply:
Find a way to construct your reference so that it does not have any characters that can be used in a state name. Perhaps use colon as delimiter (we do that elsewhere)? Or basically find a character that is forbidden in state names and use that.

Reply:
Okay

Reply:
I found a similar case of unicode chars state name in the classifier gae_models, although I checked delimiters in the code base, but the mostly ```.``` is used, but somewhere ```:``` & ```-``` is also used. Although in classifier gae_model instance it holds almost same type of instance id of the form ```{{exp_id}}.{{exp_version}}.{{utf8_encoded_state_name}}```, so I think I can go for ```.```

Reply:
Which characters are allowed in state names? I thought we had a regex for that. Please include a link to that code in your next comment (the code that validates what characters are allowed in state names) and use, for your delimiter, some character that is not in that charset.

Reply:
In other words, do what is needed so that you can have a "!= 2" here.

Reply:
https://github.com/oppia/oppia/blob/e69b57106e42f5b7f80e8556b3bd62713f4fadf2/utils.py#L575
https://github.com/oppia/oppia/blob/e69b57106e42f5b7f80e8556b3bd62713f4fadf2/assets/constants.js#L517

I am using ```:``` as a delimilter. I checked in the corresponding lines of the code and found that colon can be used.